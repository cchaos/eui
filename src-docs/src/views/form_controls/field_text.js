import React, { useState } from 'react';

import { EuiFieldText, EuiSpacer } from '../../../../src/components';
import { DisplayToggles } from './display_toggles';

export default function () {
  const [value, setValue] = useState('');

  const onChange = (e) => {
    setValue(e.target.value);
  };

  return (
    <>
      {/* DisplayToggles wrapper for Docs only */}
      <DisplayToggles canPrepend canAppend>
        <EuiFieldText
          placeholder="Placeholder text"
          value={value}
          onChange={(e) => onChange(e)}
          aria-label="Use aria labels when no actual label is in use"
        />
      </DisplayToggles>
      <EuiSpacer />
      <DisplayToggles canPrepend canAppend>
        <EuiFieldText
          placeholder="Placeholder text"
          value={value}
          onChange={(e) => onChange(e)}
          label="This text field has its own label prop"
        />
      </DisplayToggles>
    </>
  );
}
