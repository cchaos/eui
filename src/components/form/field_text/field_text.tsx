/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { InputHTMLAttributes, Ref, FunctionComponent } from 'react';
import { CommonProps } from '../../common';
import classNames from 'classnames';

import {
  EuiFormControlLayout,
  EuiFormControlLayoutProps,
} from '../form_control_layout';

import { EuiValidatableControl } from '../validatable_control';
import { EuiFormRow } from '../form_row';
import {
  EuiFormRowCommonProps,
  euiFormRowDisplayIsCompressed,
} from '../form_row/form_row';

type EuiFieldTextSupportedRowDisplays =
  | 'row'
  | 'rowCompressed'
  | 'columnCompressed';

export type EuiFieldTextProps = Omit<
  EuiFormRowCommonProps,
  'children' | 'display' | 'hasChildLabel'
> &
  InputHTMLAttributes<HTMLInputElement> &
  CommonProps & {
    icon?: EuiFormControlLayoutProps['icon'];
    isInvalid?: boolean;
    fullWidth?: boolean;
    isLoading?: boolean;
    readOnly?: boolean;
    inputRef?: Ref<HTMLInputElement>;
    placeholder?: HTMLInputElement['placeholder'];

    /**
     * Creates an input group with element(s) coming before input;
     * `string` | `ReactElement` or an array of these
     */
    prepend?: EuiFormControlLayoutProps['prepend'];

    /**
     * Creates an input group with element(s) coming after input;
     * `string` | `ReactElement` or an array of these
     */
    append?: EuiFormControlLayoutProps['append'];

    /**
     * Completely removes form control layout wrapper and ignores
     * icon, prepend, and append and all form row props;
     * Best used inside EuiFormControlLayoutDelimited
     */
    controlOnly?: boolean;

    /**
     * When `true` creates a shorter height input
     */
    compressed?: boolean;

    /**
     * Custom list of supported row displays
     */
    display?: EuiFieldTextSupportedRowDisplays;
  };

export const EuiFieldText: FunctionComponent<EuiFieldTextProps> = ({
  id,
  name,
  placeholder,
  value,
  className,
  icon,
  isInvalid,
  inputRef,
  fullWidth = false,
  isLoading,
  compressed,
  prepend,
  append,
  readOnly,
  controlOnly,
  // FormRowProps
  helpText,
  error,
  label,
  labelAppend,
  hasEmptyLabelSpace,
  describedByIds,
  display = 'row',
  ...rest
}) => {
  // Force compressed if `display` is compressed
  compressed = euiFormRowDisplayIsCompressed(display) || compressed;

  const classes = classNames('euiFieldText', className, {
    'euiFieldText--withIcon': icon,
    'euiFieldText--fullWidth': fullWidth,
    'euiFieldText--compressed': compressed,
    'euiFieldText--inGroup': prepend || append,
    'euiFieldText-isLoading': isLoading,
  });

  const control = (
    <EuiValidatableControl isInvalid={isInvalid}>
      <input
        type="text"
        id={id}
        name={name}
        placeholder={placeholder}
        className={classes}
        value={value}
        ref={inputRef}
        readOnly={readOnly}
        {...rest}
      />
    </EuiValidatableControl>
  );

  if (controlOnly) return control;

  const formControlLayout = (
    <EuiFormControlLayout
      icon={
        isInvalid
          ? {
              type: 'alert',
              color: 'danger',
              side: 'right',
            }
          : icon
      }
      fullWidth={fullWidth}
      isLoading={isLoading}
      compressed={compressed}
      readOnly={readOnly}
      prepend={prepend}
      append={append}
      inputId={id}>
      {control}
    </EuiFormControlLayout>
  );

  if (!label && !error && !helpText && !hasEmptyLabelSpace)
    return formControlLayout;

  const formRowProps = {
    helpText,
    error,
    label,
    labelAppend,
    hasEmptyLabelSpace,
    describedByIds: id ? [id] : undefined,
    display,
    fullWidth,
    isInvalid,
  };

  return <EuiFormRow {...formRowProps}>{formControlLayout}</EuiFormRow>;
};
