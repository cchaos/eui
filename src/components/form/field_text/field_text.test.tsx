/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test/required_props';

import { EuiFieldText } from './field_text';

jest.mock('../validatable_control', () => ({
  EuiValidatableControl: 'eui-validatable-control',
}));

describe('EuiFieldText', () => {
  test('is rendered', () => {
    const component = render(
      <EuiFieldText
        name="elastic"
        id="1"
        value="1"
        inputRef={() => {}}
        onChange={() => {}}
        {...requiredProps}
      />
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    test('icon is rendered', () => {
      const component = render(<EuiFieldText icon="bolt" />);

      expect(component).toMatchSnapshot();
    });

    test('isInvalid is rendered', () => {
      const component = render(<EuiFieldText isInvalid />);

      expect(component).toMatchSnapshot();
    });

    test('fullWidth is rendered', () => {
      const component = render(<EuiFieldText fullWidth />);

      expect(component).toMatchSnapshot();
    });

    test('readOnly is rendered', () => {
      const component = render(<EuiFieldText readOnly />);

      expect(component).toMatchSnapshot();
    });

    test('isLoading is rendered', () => {
      const component = render(<EuiFieldText isLoading />);

      expect(component).toMatchSnapshot();
    });

    test('compressed is rendered', () => {
      const component = render(<EuiFieldText compressed />);

      expect(component).toMatchSnapshot();
    });

    test('controlOnly is rendered', () => {
      const component = render(<EuiFieldText controlOnly />);

      expect(component).toMatchSnapshot();
    });

    test('placeholder is rendered', () => {
      const component = render(<EuiFieldText placeholder="Placeholder" />);

      expect(component).toMatchSnapshot();
    });

    test('prepend is rendered', () => {
      const component = render(<EuiFieldText prepend="prepend" />);

      expect(component).toMatchSnapshot();
    });

    test('append is rendered', () => {
      const component = render(<EuiFieldText append="append" />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('form row props', () => {
    test('are rendered', () => {
      const component = render(
        <EuiFieldText
          helpText={'helpText'}
          error={'error'}
          label={'label'}
          labelAppend={'labelAppend'}
          labelProps={requiredProps}
          hasEmptyLabelSpace={true}
          display="columnCompressed"
          fullWidth={true}
          isInvalid={true}
        />
      );

      expect(component).toMatchSnapshot();
    });
  });
});
