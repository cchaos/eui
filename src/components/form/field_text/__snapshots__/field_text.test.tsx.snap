// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiFieldText form row props are rendered 1`] = `
<div
  class="euiFormRow euiFormRow--hasEmptyLabelSpace euiFormRow--fullWidth euiFormRow--compressed euiFormRow--horizontal"
  id="generated-id-row"
>
  <div
    class="euiFormRow__labelWrapper"
  >
    <label
      aria-invalid="true"
      aria-label="aria-label"
      class="euiFormLabel testClass1 testClass2 euiFormLabel-isInvalid"
      data-test-subj="test subject string"
      for="euiFieldText_generated-id"
    >
      label
    </label>
     labelAppend
  </div>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <eui-form-control-layout
      aria-describedby="generated-id-help generated-id-error-0"
      compressed="true"
      fullwidth="true"
      inputid="euiFieldText_generated-id"
      onblur="function () {
      // Doing this to allow onBlur to be called correctly from the child input element as this component overrides it
      var onChildBlur = (0, _objects.get)(_this.props, 'children.props.onBlur');

      if (onChildBlur) {
        onChildBlur.apply(void 0, arguments);
      }

      _this.setState({
        isFocused: false
      });
    }"
      onfocus="function () {
      // Doing this to allow onFocus to be called correctly from the child input element as this component overrides it
      var onChildFocus = (0, _objects.get)(_this.props, 'children.props.onFocus');

      if (onChildFocus) {
        onChildFocus.apply(void 0, arguments);
      }

      _this.setState(function (_ref) {
        var isFocused = _ref.isFocused;

        if (!isFocused) {
          return {
            isFocused: true
          };
        } else {
          return null;
        }
      });
    }"
    >
      <eui-validatable-control
        isinvalid="true"
      >
        <input
          class="euiFieldText euiFieldText--fullWidth euiFieldText--compressed"
          id="euiFieldText_generated-id"
          type="text"
        />
      </eui-validatable-control>
    </eui-form-control-layout>
    <div
      aria-live="polite"
      class="euiFormErrorText euiFormRow__text"
      id="generated-id-error-0"
    >
      error
    </div>
    <div
      class="euiFormHelpText euiFormRow__text"
      id="generated-id-help"
    >
      helpText
    </div>
  </div>
</div>
`;

exports[`EuiFieldText is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="1"
>
  <eui-validatable-control>
    <input
      aria-label="aria-label"
      class="euiFieldText testClass1 testClass2"
      data-test-subj="test subject string"
      id="1"
      name="elastic"
      type="text"
      value="1"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props append is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control>
    <input
      class="euiFieldText"
      id="euiFieldText_generated-id"
      placeholder="append"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props compressed is rendered 1`] = `
<eui-form-control-layout
  compressed="true"
  fullwidth="false"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control>
    <input
      class="euiFieldText euiFieldText--compressed"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props controlOnly is rendered 1`] = `
<eui-validatable-control>
  <input
    class="euiFieldText"
    id="euiFieldText_generated-id"
    type="text"
  />
</eui-validatable-control>
`;

exports[`EuiFieldText props fullWidth is rendered 1`] = `
<eui-form-control-layout
  fullwidth="true"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control>
    <input
      class="euiFieldText euiFieldText--fullWidth"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props icon is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  icon="bolt"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control>
    <input
      class="euiFieldText euiFieldText--withIcon"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props isInvalid is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control
    isinvalid="true"
  >
    <input
      class="euiFieldText"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props isLoading is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
  isloading="true"
>
  <eui-validatable-control>
    <input
      class="euiFieldText euiFieldText-isLoading"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props placeholder is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
>
  <eui-validatable-control>
    <input
      class="euiFieldText"
      id="euiFieldText_generated-id"
      placeholder="Placeholder"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props prepend is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
  prepend="prepend"
>
  <eui-validatable-control>
    <input
      class="euiFieldText euiFieldText--inGroup"
      id="euiFieldText_generated-id"
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;

exports[`EuiFieldText props readOnly is rendered 1`] = `
<eui-form-control-layout
  fullwidth="false"
  inputid="euiFieldText_generated-id"
  readonly="true"
>
  <eui-validatable-control>
    <input
      class="euiFieldText"
      id="euiFieldText_generated-id"
      readonly=""
      type="text"
    />
  </eui-validatable-control>
</eui-form-control-layout>
`;
